components/causal_ts/src/tso.rs

```diff
@@ -0,0 +1,292 @@
+            match batch.used_size() {
+                None => batch_min_size,
+                Some(used_size) => {
+                    debug!("CachedTsoProvider::renew_tso_batch"; "batch before" => ?batch, "need_flush" => need_flush, "used size" => used_size);
+                    Self::calc_new_batch_size(batch.size, used_size, batch_min_size)
+                }
+            }
+        };
+
+        match pd_client.batch_get_tso(new_batch_size).await {
+            Err(err) => {
+                warn!("BatchTsoProvider::renew_tso_batch, pd_client.batch_get_tso error"; "error" => ?err, "need_flash" => need_flush);
+                if need_flush {
+                    let batch = tso_batch.write();
+                    batch.flush();
+                }
+                Err(err.into())
+            }
+            Ok(ts) => {
+                let mut batch = tso_batch.write();
+                batch.renew(new_batch_size, ts);
+                debug!("BatchTsoProvider::renew_tso_batch"; "batch renew" => ?batch, "ts" => ?ts);
+                Ok(())
+            }
+        }
+    }
+
+    fn calc_new_batch_size(batch_size: u32, used_size: u32, batch_min_size: u32) -> u32 {
+        if used_size > batch_size * 3 / 4 {
+            // Enlarge to double if used more than 3/4.
+            std::cmp::min(batch_size << 1, TSO_BATCH_MAX_SIZE)
+        } else if used_size < batch_size / 4 {
+            // Shrink to half if used less than 1/4.
+            std::cmp::max(batch_size >> 1, batch_min_size)
+        } else {
+            batch_size
+        }
+    }
+
+    async fn init(&self) -> Result<()> {
+        self.renew_tso_batch(true).await?;
+
+        let pd_client = self.pd_client.clone();
+        let tso_batch = self.batch.clone();
+        let batch_min_size = self.batch_min_size;
+        let task = move || {
+            let pd_client = pd_client.clone();
+            let tso_batch = tso_batch.clone();
+            async move {
+                let _ = Self::renew_tso_batch_internal(pd_client, tso_batch, batch_min_size, false)
+                    .await;
+            }
+        };
+
+        // Duration::ZERO means never renew automatically. For test purpose ONLY.
+        if self.renew_interval > Duration::ZERO {
+            self.renew_worker
+                .spawn_interval_async_task(self.renew_interval, task);
+        }
+        Ok(())
+    }
+
+    // Get current batch_size, for test purpose.
+    pub fn batch_size(&self) -> u32 {
+        self.batch.read().size
+    }
+}
+
+impl CausalTsProvider for BatchTsoProvider {
+    fn get_ts(&self) -> Result<TimeStamp> {
+        let batch = self.batch.read();
+        match batch.pop() {
+            Some(ts) => {
+                trace!("BatchTsoProvider::get_ts: {:?}", ts);
+                Ok(ts)
+            }
+            None => {
+                error!("BatchTsoProvider::get_ts, batch used up"; "batch.size" => batch.size);
+                Err(Error::TsoBatchUsedUp(batch.size))
```

Comment:
when it's used up, renew immediately?

Reply:
The batch size is expected to be always big enough, and used up should not happen.
The default initial batch size is set according to benchmark.
Moreover, the initial batch size is configurable, so in case there is a rare scenario, user can still set a bigger initial batch size to avoid the used up.

Renew immediately may need carefully handle burst requests in an instant, to avoid impacting the TSO service, and will make the QPS slow down. Setting to a bigger initial batch size would be a better solution.

Reply:
why does renewing immediately impact the TSO service? tso service is light-weight which can serve high enough QPS.
And if you don't renew, the request may retry which slows down QPS as well.

Reply:
En...my main idea here is that, the batch size should be always enough. If it is not (it should be very rare), user should adjust initial batch size in configuration. And also maybe we need to optimize the batch size algorithm.
So I think we don't need the renew here.

Reply:
Batch renew on used up is added. Also add some metrics, for further analysis & optimization.
PTAL, thanks~